### PR TITLE
Repeatedly recalculate states

### LINKED ISSUE #67: Handle block triggering multiple state recalcs
# Issue
Repo currently doesn't handle the case in which a block's parent spans multiple state recalc points. For example if the parent of slot 258 was at slot 65. Then processing block at 258 would trigger a  state recalc for 128, 192, and 256. The mechanism to solve this is described in the spec:

> Repeat while `slot - last_state_recalc >= CYCLE_LENGTH`:

# Proposed implementation
Add a `while` loop around the `initialize_new_cycle` part of `compute_state_transition` instead of the current `if`

See here: https://github.com/ethereum/beacon_chain/blob/master/beacon_chain/state/state_transition.py#L333

### PR DESCRIPTION
**Pull request by hwwhww**
### What's wrong
Fixed #67

### How was it fixed?
* Added a `get_recalculated_states` function for easily testing.
* Now there's only one test case that Danny mentioned.

### CODE REVIEW

File: tests/state/helpers.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def get_pseudo_chain(length):
         blocks.append(
             Block(
                 slot_number=slot,
-                parent_hash=blocks[slot-1].hash if slot > 0 else b'00'*32
+                parent_hash=blocks[slot-1].hash if slot > 0 else b'\x00'*32
```
Comment (jannikluhn) [inline tests/state/helpers.py]: `ZERO_HASH32`?

Also formatting: `blocks[slot-1]` -> `blocks[slot - 1]`

File: tests/state/test_state_transition.py [hunk 2]
```diff
@@ -29,3 +38,89 @@ def test_validate_attestation_bitfield():
...
 def test_validate_attestation_aggregate_sig():
     pass
+
+
+@pytest.mark.parametrize(
+    (
+        'cycle_length,last_state_recalc,'
+        'last_justified_slot,justified_streak,last_finalized_slot,'
+        'crystallized_state_total_deposits,block_vote_cache_total_deposits,'
+        'result_last_state_recalc,result_justified_streak, result_last_finalized_slot'
```
Comment (jannikluhn) [inline tests/state/test_state_transition.py]: One argument per line would be easier to read.

File: tests/state/test_state_transition.py [hunk 3]
```diff
@@ -29,3 +38,89 @@ def test_validate_attestation_bitfield():
...
+    ),
+    [
+        (64, 0, 0, 0, 0, 3, 2, 64, 0+64, 0),  # 2/3 attestations
```
Comment (jannikluhn) [inline tests/state/test_state_transition.py]: formatting: `0+64` -> `0 + 64` (same in following lines)

File: beacon_chain/state/state_transition.py [hunk 4]
```diff
@@ -348,3 +347,18 @@ def compute_state_transition(parent_state: Tuple[CrystallizedState, ActiveState]
     )
 
     return crystallized_state, active_state
+
+
+def get_recalculated_states(
```
Comment (jannikluhn) [inline beacon_chain/state/state_transition.py]: Maybe the name should be more specific as this seems to just handle cycle transitions. Maybe something like `compute_cycle_transitions`?

**Comment by djrtwo** [comment]
Looks good to me after @jannikluhn's proposed changes

File: tests/state/test_state_transition.py (hunk 3 above)
Comment (hwwhww) [reply tests/state/test_state_transition.py]: I find it's harder to read:
```python
    [
        (64, 0, 0, 0, 0, 3, 2, 64, 0 + 64, 0),  # 2/3 attestations
        (64, 0, 0, 0, 0, 3, 1, 64, 0 + 0, 0),   # 1/3 attestations
        (64, 128, 128, 64, 0, 3, 2, 128 + 64, 64 + 64, 127 - 64 - 1),  # 2/3 attestations, last_finalized_slot = slot - cycle_length - 1
        (64, 128, 128, 128, 128, 3, 2, 128 + 64, 128 + 64, 128),  # 2/3 attestations, last_finalized_slot = last_finalized_slot
    ],
```
Not the same case, but similar to something like https://www.python.org/dev/peps/pep-0008/#other-recommendations
`hypot2 = x*x + y*y` is better than `hypot2 = x * x + y * y` sometimes.

**Pull request merged**